File: backend/app/schemas/social.py
from datetime import datetime
from typing import Optional, Literal

from pydantic import BaseModel, Field, ConfigDict


Platform = Literal["pulseforge", "linkedin", "twitter", "facebook", "reddit"]
PostStatus = Literal["pending", "scheduled", "posting", "posted", "failed", "cancelled"]


class SocialAccountBase(BaseModel):
    platform: Platform
    account_name: str = Field(min_length=1, max_length=100)
    access_token: Optional[str] = None
    config: Optional[dict] = None
    is_active: bool = True


class SocialAccountCreate(SocialAccountBase):
    pass


class SocialAccountUpdate(BaseModel):
    account_name: Optional[str] = None
    access_token: Optional[str] = None
    config: Optional[dict] = None
    is_active: Optional[bool] = None


class SocialAccountOut(SocialAccountBase):
    id: int
    user_id: int
    token_expires_at: Optional[datetime] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class SocialPostBase(BaseModel):
    account_id: int
    title: Optional[str] = Field(default=None, max_length=300)
    summary: Optional[str] = None
    image_url: Optional[str] = None
    external_url: Optional[str] = None
    hashtags: Optional[list[str]] = None
    scheduled_at: Optional[datetime] = None


class SocialPostCreate(SocialPostBase):
    pass


class SocialPostUpdate(BaseModel):
    title: Optional[str] = None
    summary: Optional[str] = None
    image_url: Optional[str] = None
    external_url: Optional[str] = None
    hashtags: Optional[list[str]] = None
    scheduled_at: Optional[datetime] = None


class SocialPostOut(SocialPostBase):
    id: int
    status: PostStatus
    platform_post_id: Optional[str] = None
    posted_at: Optional[datetime] = None
    engagement: Optional[dict] = None
    error_message: Optional[str] = None
    retry_count: int = 0
    created_at: datetime
    updated_at: datetime
    platform: Optional[str] = None
    account_name: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


class SocialPostSendNow(BaseModel):
    post_id: int

File: frontend/src/api/social.ts
import http from './http'

export type Platform = 'pulseforge' | 'linkedin' | 'twitter' | 'facebook' | 'reddit'
export type PostStatus = 'pending' | 'scheduled' | 'posting' | 'posted' | 'failed' | 'cancelled'

export interface SocialAccount {
  id: number
  user_id: number
  platform: Platform
  account_name: string
  access_token?: string
  token_expires_at?: string | null
  config?: Record<string, any> | null
  is_active: boolean
  created_at: string
}

export interface SocialAccountCreate {
  platform: Platform
  account_name: string
  access_token?: string
  config?: Record<string, any>
  is_active?: boolean
}

export interface SocialAccountUpdate {
  account_name?: string
  access_token?: string
  config?: Record<string, any>
  is_active?: boolean
}

export interface SocialPost {
  id: number
  account_id: number
  title?: string
  summary?: string
  image_url?: string
  external_url?: string
  hashtags?: string[] | null
  scheduled_at?: string | null
  posted_at?: string | null
  platform_post_id?: string | null
  status: PostStatus
  engagement?: Record<string, any> | null
  error_message?: string | null
  retry_count: number
  created_at: string
  updated_at: string
  platform?: string
  account_name?: string
}

export interface SocialPostCreate {
  account_id: number
  title?: string
  summary?: string
  image_url?: string
  external_url?: string
  hashtags?: string[]
  scheduled_at?: string
}

export interface SocialPostUpdate {
  title?: string
  summary?: string
  image_url?: string
  external_url?: string
  hashtags?: string[]
  scheduled_at?: string
}

export function listAccounts() {
  return http.get<any, SocialAccount[]>('/social/accounts')
}

export function createAccount(payload: SocialAccountCreate) {
  return http.post<any, SocialAccount>('/social/accounts', payload)
}

export function updateAccount(id: number, payload: SocialAccountUpdate) {
  return http.patch<any, SocialAccount>(`/social/accounts/${id}`, payload)
}

export function deleteAccount(id: number) {
  return http.delete(`/social/accounts/${id}`)
}

export function listPosts(params?: { status?: string; account_id?: number }) {
  return http.get<any, SocialPost[]>('/social/posts', { params })
}

export function createPost(payload: SocialPostCreate) {
  return http.post<any, SocialPost>('/social/posts', payload)
}

export function updatePost(id: number, payload: SocialPostUpdate) {
  return http.patch<any, SocialPost>(`/social/posts/${id}`, payload)
}

export function deletePost(id: number) {
  return http.delete(`/social/posts/${id}`)
}

export function sendPostNow(id: number) {
  return http.post<any, SocialPost>(`/social/posts/${id}/send-now`, {})
}
